### PR TITLE
[KYUUBI #2474] [Improvement] Add FilteredShowFunctionsCommand to Authz module

### PR DESCRIPTION
**Pull request by zhaomin1423**
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
please add some tests with the system default functions

**Comment by zhaomin1423** [comment]
> please add some tests with the system default functions

added, but CI/TEST(11,3.2) is failed, I will examine it later.

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 1]
```diff
@@ -273,6 +273,30 @@ abstract class RangerSparkExtensionSuite extends KyuubiFunSuite with SparkSessio
...
+      doAs("bob", assert(sql(s"show user functions $default.$function1").collect().length == 0))
+
+      doAs("admin", sql(s"CREATE DATABASE IF NOT EXISTS $db3"))
+      doAs("admin", sql(s"CREATE FUNCTION $db3.$function1 AS 'Function1'"))
+
+      doAs("admin", assert(sql(s"show user functions $db3.$function1").collect().length == 1))
+      doAs("bob", assert(sql(s"show user functions $db3.$function1").collect().length == 0))
+
+      doAs("admin", assert(sql(s"show system functions").collect().length > 0))
+      doAs("bob", assert(sql(s"show system functions").collect().length === 0))
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:293]: the builtin functions shall have open ACLs

Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: updated, the builtin functions do't check.

**Comment by MLikeWater** [comment]
@zhaomin1423 FilteredShowFunctionsCommand just returns `show functions` result, but if the query using udf functions, such as `select testUDF (id) from the test`, this also needs to be implemented through ranger control.
```
# Ranger does not authorize users to access the UDF
0: jdbc:hive2://xx.xx.xx.xxx:10011/default> show functions
+-----------+
| function  |
+-----------+
+-----------+
0: jdbc:hive2://xx.xx.xx.xxx:10011/default> SELECT testUDF(id) FROM test;
+-------------------+
| shdw.testUDF(id)  |
+-------------------+
| 1                 |
+-------------------+

```

**Comment by zhaomin1423** [comment]
> @zhaomin1423 FilteredShowFunctionsCommand just returns `show functions` result, but if the query using udf functions, such as `select testUDF (id) from the test`, this also needs to be implemented through ranger control.
> 
> ```
> # Ranger does not authorize users to access the UDF
> 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> show functions
> +-----------+
> | function  |
> +-----------+
> +-----------+
> 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> SELECT testUDF(id) FROM test;
> +-------------------+
> | shdw.testUDF(id)  |
> +-------------------+
> | 1                 |
> +-------------------+
> ```

Thanks, it may need other command to handle it.

**Comment by yaooqinn** [comment]
please resolve the conflicts and address comments

**Comment by zhaomin1423** [comment]
> @zhaomin1423 FilteredShowFunctionsCommand just returns `show functions` result, but if the query using udf functions, such as `select testUDF (id) from the test`, this also needs to be implemented through ranger control.
> 
> ```
> # Ranger does not authorize users to access the UDF
> 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> show functions
> +-----------+
> | function  |
> +-----------+
> +-----------+
> 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> SELECT testUDF(id) FROM test;
> +-------------------+
> | shdw.testUDF(id)  |
> +-------------------+
> | 1                 |
> +-------------------+
> ```

Could we resolve the case in the pr？ @yaooqinn

**Comment by yaooqinn** [comment]
what are the behaviors of hive and presto？

**Comment by yaooqinn** [comment]
> > @zhaomin1423 FilteredShowFunctionsCommand just returns `show functions` result, but if the query using udf functions, such as `select testUDF (id) from the test`, this also needs to be implemented through ranger control.
> > ```
> > # Ranger does not authorize users to access the UDF
> > 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> show functions
> > +-----------+
> > | function  |
> > +-----------+
> > +-----------+
> > 0: jdbc:hive2://xx.xx.xx.xxx:10011/default> SELECT testUDF(id) FROM test;
> > +-------------------+
> > | shdw.testUDF(id)  |
> > +-------------------+
> > | 1                 |
> > +-------------------+
> > ```
> 
> Could we resolve the case in the pr？ @yaooqinn

no

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleReplaceShowObjectCommands.scala [hunk 2]
```diff
@@ -81,3 +83,21 @@ abstract class FilteredShowObjectCommand(delegated: RunnableCommand)
...
 }
+
+case class FilteredShowFunctionsCommand(delegated: RunnableCommand)
+  extends FilteredShowObjectCommand(delegated) with WithInternalChild {
+
+  override protected def isAllowed(r: Row, ugi: UserGroupInformation): Boolean = {
+    val functionName = r.getString(0)
+    val items = functionName.split("\\.", 2)
+    // the system functions return true
+    if (items.length == 1) {
```
Comment (zhaomin1423) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleReplaceShowObjectCommands.scala:94]: I don't know if it's ok？

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**